File: 496.go
package leetcode

func nextGreaterElement(nums1 []int, nums2 []int) []int {
	mp := map[int]int{}
	for idx, val := range nums1 {
		mp[val] = idx
	}
	s := []int{}
	ans := make([]int, len(nums1))
	for i := len(nums2) - 1; i >= 0; i-- {
		for len(s) > 0 && s[len(s)-1] <= nums2[i] {
			s = s[:len(s)-1]
		}
		if idx, ok := mp[nums2[i]]; ok {
			if len(s) > 0 {
				ans[idx] = s[len(s)-1]
			} else {
				ans[idx] = -1
			}
		}
		s = append(s, nums2[i])
	}
	return ans
}

File: 496.py
from typing import List

class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        mp = {val:idx for idx, val in enumerate(nums1)}
        s = []
        ans = [0] * len(nums1)
        for val in nums2[::-1]:
            while s and s[-1] <= val:
                s.pop()
            bigger_num = s[-1] if s else -1
            if val in mp:
                ans[mp[val]] = bigger_num
            s.append(val)
        return ans